moveit_planners/stomp/include/stomp_moveit/cost_functions.hpp

```diff
@@ -0,0 +1,245 @@
+      bool found_invalid_state = false;
+      while (!found_invalid_state && interpolation_fraction < 1.0)
+      {
+        Eigen::VectorXd sample_vec = (1 - interpolation_fraction) * current + interpolation_fraction * next;
+
+        found_invalid_state = !state_validator_fn(sample_vec);
+        interpolation_fraction += interpolation_step;
+      }
+
+      if (found_invalid_state)
+      {
+        // Apply weighted penalties -> This trajectory is definitely invalid
+        costs(timestep) = (1 - interpolation_fraction) * penalty;
+        costs(timestep + 1) = interpolation_fraction * penalty;
+        validity = false;
+
+        // Open new invalid window when this is the first detected invalid state in a group
+        if (!in_invalid_window)
+        {
+          invalid_windows.emplace_back(timestep, values.cols());
+          in_invalid_window = true;
+        }
+      }
+      else
+      {
+        // Close current invalid window if the current state is valid
+        if (in_invalid_window)
+        {
+          invalid_windows.back().second = timestep - 1;
+          in_invalid_window = false;
+        }
+      }
+    }
+
+    // Smooth out cost of invalid segments using a gaussian
+    // The standard deviation is picked so that neighboring states
+    // before and after the violation are penalized as well.
+    for (const auto& [start, end] : invalid_windows)
+    {
+      const double window_size = static_cast<double>(end - start) + 1;
+      const double sigma = window_size / 5.0;
+      const double mu = 0.5 * (start + end);
+      for (auto j = std::max(0l, start - static_cast<long>(sigma));
+           j <= std::min(values.cols() - 1, end + static_cast<long>(sigma)); ++j)
+      {
+        costs(j) +=
+            std::exp(-std::pow(j - mu, 2) / (2 * std::pow(sigma, 2))) / (sigma * std::sqrt(2 * mu)) * window_size;
+      }
+    }
+
+    return true;
+  };
+
+  return cost_fn;
+}
+
+/**
+ * Creates a cost function for binary collisions of group states in the planning scene.
+ * This function uses a StateValidatorFn for computing smooth penalty costs from binary
+ * collision checks using get_cost_function_from_state_validator().
+ *
+ * @param planning_scene    The planning scene instance to use for collision checking
+ * @param group             The group to use for computing link transforms from joint positions
+ * @param collision_penalty The penalty cost value applied to colliding states
+ *
+ * @return                  Cost function that computes smooth costs for colliding path segments
+ */
+CostFn get_collision_cost_function(const std::shared_ptr<const planning_scene::PlanningScene>& planning_scene,
+                                   const moveit::core::JointModelGroup* group, double collision_penalty)
+{
+  const auto& joints = group ? group->getActiveJointModels() : planning_scene->getRobotModel()->getActiveJointModels();
```

Comment:
why would it matter? If a group is specified, the group is used. Otherwise, the robot model is used. The group is technically a filter (of joints) on the robot model.